File: src/cnf/navigation/astar/heuristics.py

import numpy as np

from typing import List

from cnf import CrystalNormalForm
from ...utils.pdd import pdd_for_cnfs, pdd_amd_for_cnfs
from ...lattice.permutations import VonormPermutation, CONORM_PERMUTATION_TO_VONORM_PERMUTATION, ZERO_CONORM_SETS_TO_PERMUTATIONS_TO_UNIMOD_MATS
from ...lattice.vonorm_unimodular import VonormPermutationMatrix
from ...lattice.voronoi.vonorm_list import VonormList, VONORM_TO_DOT_PRODUCTS
from ...motif.mnf_constructor import sort_motif_coord_arr, move_coords_into_bounds, get_mnf_strs_from_coord_mats

def pdd_heuristic(cnf: tuple, goals: list[CrystalNormalForm]) -> float:
    xi = goals[0].xi
    delta = goals[0].delta
    els = goals[0].elements

    pt = CrystalNormalForm.from_tuple(cnf, els, xi, delta)
    dists = [pdd_for_cnfs(pt, g, k=20) for g in goals]
    return (min(dists) * 100) ** 2

def pdd_and_manhattan(cnf: tuple, goals: list[CrystalNormalForm]) -> float:
    xi = goals[0].xi
    delta = goals[0].delta
    els = goals[0].elements

    pt = CrystalNormalForm.from_tuple(cnf, els, xi, delta)
    dists = [pdd_for_cnfs(pt, g, k=20) for g in goals]
    return 10000 * min(dists) + manhattan_distance(cnf, goals)

def pdd_amd_heuristic(cnf: tuple, goals: list[CrystalNormalForm]) -> float:
    xi = goals[0].xi
    delta = goals[0].delta
    els = goals[0].elements

    pt = CrystalNormalForm.from_tuple(cnf, els, xi, delta)
    dists = [pdd_amd_for_cnfs(pt, g, k=20) for g in goals]
    return (min(dists) * 100) ** 3 


def manhattan_distance(cnf: tuple, goals: list[CrystalNormalForm]) -> float:
    manhattan_dist = float('inf')
    current_coords = np.array(cnf)

    for goal in goals:
        goal_coords = np.array(goal.coords)
        curr_dist = np.sum(np.abs(current_coords - goal_coords))
        manhattan_dist = min(manhattan_dist, curr_dist)

    return manhattan_dist * 2

def manhattan_dist_cnfs(cnf: CrystalNormalForm, cnf2: CrystalNormalForm) -> float:
    return np.sum(np.abs(np.array(cnf.coords) - np.array(cnf2.coords)))

def squared_euclidean_heuristic(cnf: tuple, goals: List[CrystalNormalForm]) -> float:

    min_dist_sq = float('inf')

    for goal in goals:
        goal_coords = np.array(goal.coords)
        dist_sq = np.sum((np.array(cnf) - goal_coords) ** 2)
        min_dist_sq = min(min_dist_sq, dist_sq)

    return min_dist_sq


# ---------------------------------------------------------------------------
# Unimodular-aware Manhattan heuristic
# ---------------------------------------------------------------------------
#
# The plain Manhattan heuristic ignores the fact that the canonical ordering
# of vonorms (and motif coordinates) can change across Voronoi-class
# boundaries.  Two CNFs that are one graph step apart may look very far
# apart in Manhattan distance because they sit in different canonical
# orderings.
#
# This heuristic pre-computes all 168 vonorm-permutation variants of each
# goal CNF (permuted vonorms + unimodular-transformed & re-sorted motif).
# During search it returns the *minimum* Manhattan distance over all
# variants, giving a much tighter estimate of the true graph distance.
# ---------------------------------------------------------------------------

def _precompute_goal_variants(goal: CrystalNormalForm):
    """Pre-compute all vonorm-permuted + motif-transformed variants of a goal.

    For each of the 168 vonorm permutations we:
      1. Permute the 7 vonorms.
      2. Apply the corresponding inverse unimodular matrix to the motif
         coordinate matrix (integer arithmetic, mod delta).
      3. Try every valid origin shift (one per atom of the first element).
      4. Sort atoms lexicographically by (element, x, y, z).
      5. Flatten to the CNF tuple layout: vonorms ++ motif_coords.

    Returns:
        np.ndarray of shape (N_variants, tuple_len) containing all goal
        representations, ready for vectorised Manhattan-distance computation.
    """
    vonorms = np.array(goal.coords[:7])
    delta = goal.delta
    elements = goal.elements

    # Reconstruct motif coord matrix (3, N-1) from the MNF tuple
    mnf_tuple = goal.coords[7:]
    n_stored = len(mnf_tuple) // 3   # atoms stored (excludes origin)
    n_atoms = n_stored + 1
    stored_coords = np.array(mnf_tuple, dtype=np.int64).reshape(n_stored, 3)  # (N-1, 3)
    # Full coord matrix including origin at (0,0,0)
    full_coords = np.vstack([np.zeros((1, 3), dtype=np.int64), stored_coords])  # (N, 3)
    coord_matrix = full_coords.T  # (3, N)

    # Atom labels for sorting (group by element)
    atom_labels = []
    el_num = 0
    prev_el = elements[0]
    for el in elements:
        if el != prev_el:
            el_num += 1
            prev_el = el
        atom_labels.append(el_num)
    atom_labels = np.array(atom_labels)

    # Number of origin-shift candidates (atoms of the first element)
    num_origin_atoms = sum(1 for el in elements if el == elements[0])

    all_perms = VonormPermutation.all_vonorm_perm_tuples()

    variants = []
    for perm_tuple in all_perms:
        perm_arr = np.array(perm_tuple)
        permuted_vonorms = vonorms[perm_arr]

        # Unimodular matrix from positions [1:4] of the vonorm permutation
        U = VonormPermutationMatrix.from_vector_idxs(perm_tuple[1:4]).matrix
        U_inv = np.round(np.linalg.inv(U)).astype(np.int64)

        # Transform motif coords: new_coords = U_inv @ coord_matrix
        transformed = (U_inv @ coord_matrix)  # (3, N)
        transformed = np.mod(transformed, delta).astype(np.int64)

        # Try each origin shift
        for origin_idx in range(num_origin_atoms):
            shift = -transformed[:, origin_idx]  # (3,)
            shifted = transformed + shift[:, np.newaxis]
            shifted = np.mod(shifted, delta).astype(np.int64)

            # Sort by (element, x, y, z)
            sorted_cm = sort_motif_coord_arr(shifted, atom_labels)  # (3, N)

            # Extract MNF-style tuple: skip first atom (origin), flatten
            motif_tuple = sorted_cm.T[1:].flatten()

            variant = np.concatenate([permuted_vonorms, motif_tuple])
            variants.append(variant)

    return np.array(variants, dtype=np.int64)


def _precompute_goal_variants_full(goal: CrystalNormalForm, partial: bool = False):
    """Pre-compute goal variants using unimodular matrices across ALL coforms.

    Unlike ``_precompute_goal_variants`` which derives a single 3x3 matrix
    from each of 168 vonorm permutations, this version iterates over every
    conorm zero-set and every permissible conorm permutation therein,
    retrieving unimodular matrices that realise each permutation.

    This is necessary because a lattice neighbor may sit in a different
    Voronoi class than the goal, so we need transformations from all classes.

    Args:
        goal: The goal CNF to generate variants for.
        partial: If True, use only one matrix per (zero_set, conorm_perm)
            instead of all matrices. Much faster with some loss of accuracy.
    """
    vonorms = np.array(goal.coords[:7])
    delta = goal.delta
    elements = goal.elements

    # Reconstruct motif coord matrix (3, N) from the MNF tuple
    mnf_tuple = goal.coords[7:]
    n_stored = len(mnf_tuple) // 3
    stored_coords = np.array(mnf_tuple, dtype=np.int64).reshape(n_stored, 3)
    full_coords = np.vstack([np.zeros((1, 3), dtype=np.int64), stored_coords])
    coord_matrix = full_coords.T  # (3, N)

    # Atom labels for sorting
    atom_labels = []
    el_num = 0
    prev_el = elements[0]
    for el in elements:
        if el != prev_el:
            el_num += 1
            prev_el = el
        atom_labels.append(el_num)
    atom_labels = np.array(atom_labels)

    num_origin_atoms = sum(1 for el in elements if el == elements[0])

    # Iterate over ALL zero-sets (all coforms) so that cross-boundary
    # neighbors are covered regardless of Voronoi class.
    seen = set()
    variants = []
    for zero_set, perm_to_mats in ZERO_CONORM_SETS_TO_PERMUTATIONS_TO_UNIMOD_MATS.items():
        for conorm_perm, mat_list in perm_to_mats.items():
            vonorm_perm = CONORM_PERMUTATION_TO_VONORM_PERMUTATION[conorm_perm]
            perm_arr = np.array(vonorm_perm)
            permuted_vonorms = vonorms[perm_arr]

            mats_to_use = [mat_list[0]] if partial else mat_list
            for mat_tuple in mats_to_use:
                # Deduplicate by matrix identity
                mat_key = mat_tuple.tuple
                if mat_key in seen:
                    continue
                seen.add(mat_key)

                U = mat_tuple.matrix
                U_inv = np.round(np.linalg.inv(U)).astype(np.int64)

                transformed = np.mod(U_inv @ coord_matrix, delta).astype(np.int64)

                for origin_idx in range(num_origin_atoms):
                    shift = -transformed[:, origin_idx]
                    shifted = np.mod(transformed + shift[:, np.newaxis], delta).astype(np.int64)

                    sorted_cm = sort_motif_coord_arr(shifted, atom_labels)
                    motif_tuple = sorted_cm.T[1:].flatten()

                    variant = np.concatenate([permuted_vonorms, motif_tuple])
                    variants.append(variant)

    return np.unique(np.array(variants, dtype=np.int64), axis=0)


def make_heuristic(mode: str, weight: float = 0.5):
    """Factory function for A* heuristics.

    Args:
        mode: One of "manhattan", "unimodular_light", "unimodular_partial",
            "unimodular_full".
        weight: Weight for unimodular heuristics.

    Returns:
        A callable heuristic function.
    """
    if mode == "manhattan":
        return manhattan_distance
    elif mode == "unimodular_light":
        return UnimodularManhattanHeuristic(weight=weight, full=False, partial=False)
    elif mode == "unimodular_partial":
        return UnimodularManhattanHeuristic(weight=weight, full=False, partial=True)
    elif mode == "unimodular_full":
        return UnimodularManhattanHeuristic(weight=weight, full=True, partial=False)
    else:
        raise ValueError(f"Unknown heuristic mode: {mode}")


class UnimodularManhattanHeuristic:
    """Manhattan heuristic that accounts for vonorm reorderings.

    On first call, pre-computes all permuted variants of the goal CNFs.
    Subsequent calls are a single vectorised Manhattan distance computation.

    Usage::

        heuristic = UnimodularManhattanHeuristic()
        state = astar_pathfind(starts, goals, heuristic=heuristic)
    """

    def __init__(self, weight: float = 0.5, full: bool = True, partial: bool = False):
        self._goal_variants = None   # (N_variants, tuple_len) array
        self._goals_id = None        # identity check for cache invalidation
        self.weight = weight
        self.full = full
        self.partial = partial

    def _ensure_precomputed(self, goals: list[CrystalNormalForm]):
        goals_id = id(goals)
        if self._goal_variants is not None and self._goals_id == goals_id:
            return
        if self.full:
            precompute = lambda g: _precompute_goal_variants_full(g, partial=self.partial)
        else:
            precompute = _precompute_goal_variants
        variant_list = []
        for goal in goals:
            variant_list.append(precompute(goal))
        self._goal_variants = np.vstack(variant_list)
        self._goals_id = goals_id

    def __call__(self, cnf: tuple, goals: list[CrystalNormalForm]) -> float:
        self._ensure_precomputed(goals)
        current = np.array(cnf, dtype=np.int64)
        dists = np.sum(np.abs(self._goal_variants - current), axis=1)
        return float(np.min(dists)) * self.weight

File: src/cnf/rust_cnf/src/heuristics.rs
/// Unimodular Manhattan heuristic for A* pathfinding
///
/// Pre-computes all unimodular-transformed variants of goal CNFs, then returns
/// the minimum Manhattan (L1) distance over all variants. This gives a much
/// tighter estimate than plain Manhattan distance, which overestimates at
/// Voronoi class boundaries where canonical orderings change.

use std::collections::HashSet;
use crate::linalg::{mat_inv, mat_to_flat};
use crate::mnf::compute_atom_labels;
use crate::permutations::PERMUTATIONS;

// =============================================================================
// Manhattan Heuristic
// =============================================================================

/// Compute Manhattan distance (L1) heuristic between two CNF states.
///
/// Returns sum of absolute differences multiplied by 10 (matches Python implementation).
pub fn manhattan_heuristic(
    vonorms1: &[i32], coords1: &[i32],
    vonorms2: &[i32], coords2: &[i32],
) -> f64 {
    debug_assert_eq!(coords1.len(), coords2.len());
    debug_assert_eq!(vonorms1.len(), vonorms2.len());

    let vonorm_dist: i32 = vonorms1.iter()
        .zip(vonorms2.iter())
        .map(|(&v1, &v2)| (v1 - v2).abs())
        .sum();

    let coord_dist: i32 = coords1.iter()
        .zip(coords2.iter())
        .map(|(&c1, &c2)| (c1 - c2).abs())
        .sum();

    (vonorm_dist + coord_dist) as f64 * 10.0
}

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/// Heuristic mode for A* search
#[derive(Clone, Copy, Debug)]
pub enum HeuristicMode {
    /// Plain Manhattan distance (L1) with 10x scaling
    Manhattan,
    /// 168 vonorm-permutation-derived matrices (one per permutation)
    UnimodularLight,
    /// One matrix per (zero_set, conorm_perm) across all coforms
    UnimodularPartial,
    /// All matrices across all coforms
    UnimodularFull,
}

impl HeuristicMode {
    pub fn from_str(s: &str) -> Self {
        match s {
            "manhattan" => HeuristicMode::Manhattan,
            "unimodular_light" => HeuristicMode::UnimodularLight,
            "unimodular_partial" => HeuristicMode::UnimodularPartial,
            "unimodular_full" => HeuristicMode::UnimodularFull,
            _ => panic!("Unknown heuristic mode: '{}'. Expected one of: manhattan, unimodular_light, unimodular_partial, unimodular_full", s),
        }
    }
}

/// Pre-computed goal variants for unimodular manhattan heuristic
pub struct GoalVariants {
    /// Each row is a flattened [vonorms(7) ++ motif_coords((n_atoms-1)*3)] variant
    pub variants: Vec<Vec<i32>>,
    pub weight: f64,
}

// ---------------------------------------------------------------------------
// Voronoi index -> superbasis column vector mapping (for light mode)
// ---------------------------------------------------------------------------

/// Column vectors mapping Voronoi vector index to the 3D column of the
/// unimodular matrix. Used in light mode to construct matrices from vonorm
/// permutation indices [1:4].
///
/// Matches Python's VORONOI_IDX_TO_COLUMN in vonorm_unimodular.py
const VORONOI_IDX_TO_COLUMN: [[i32; 3]; 7] = [
    [-1, -1, -1], // idx 0: v3 = -(v0+v1+v2)
    [ 1,  0,  0], // idx 1: v0
    [ 0,  1,  0], // idx 2: v1
    [ 0,  0,  1], // idx 3: v2
    [ 0, -1, -1], // idx 4: v0+v1
    [-1,  0, -1], // idx 5: v0+v2
    [-1, -1,  0], // idx 6: v0+v3
];

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------

/// Build a 3x3 unimodular matrix from 3 Voronoi vector indices.
/// Corresponds to Python's `VonormPermutationMatrix.from_vector_idxs(perm[1:4])`.
fn matrix_from_voronoi_idxs(idxs: [usize; 3]) -> [[i32; 3]; 3] {
    let c0 = VORONOI_IDX_TO_COLUMN[idxs[0]];
    let c1 = VORONOI_IDX_TO_COLUMN[idxs[1]];
    let c2 = VORONOI_IDX_TO_COLUMN[idxs[2]];
    // Matrix columns -> row-major [row][col]
    [
        [c0[0], c1[0], c2[0]],
        [c0[1], c1[1], c2[1]],
        [c0[2], c1[2], c2[2]],
    ]
}

/// Positive modulo (Euclidean remainder), result in [0, m)
#[inline]
fn pos_mod(a: i32, m: i32) -> i32 {
    ((a % m) + m) % m
}

/// Transform coordinates: result[atom][dim] = (U_inv @ coords[atom]) mod delta
///
/// `full_coords` is flat [x0,y0,z0, x1,y1,z1, ...] for n_atoms atoms.
fn transform_coords(
    u_inv: &[[i32; 3]; 3],
    full_coords: &[i32],
    n_atoms: usize,
    delta: i32,
) -> Vec<i32> {
    let mut result = vec![0i32; n_atoms * 3];
    for j in 0..n_atoms {
        let off = j * 3;
        let x = full_coords[off];
        let y = full_coords[off + 1];
        let z = full_coords[off + 2];
        result[off]     = pos_mod(u_inv[0][0] * x + u_inv[0][1] * y + u_inv[0][2] * z, delta);
        result[off + 1] = pos_mod(u_inv[1][0] * x + u_inv[1][1] * y + u_inv[1][2] * z, delta);
        result[off + 2] = pos_mod(u_inv[2][0] * x + u_inv[2][1] * y + u_inv[2][2] * z, delta);
    }
    result
}

/// Sort atoms by (label, x, y, z) lexicographically.
/// `coords` is flat [x0,y0,z0, x1,y1,z1, ...].
fn sort_atoms(coords: &[i32], atom_labels: &[usize], n_atoms: usize) -> Vec<i32> {
    let mut indices: Vec<usize> = (0..n_atoms).collect();
    indices.sort_by(|&a, &b| {
        atom_labels[a]
            .cmp(&atom_labels[b])
            .then_with(|| coords[a * 3].cmp(&coords[b * 3]))
            .then_with(|| coords[a * 3 + 1].cmp(&coords[b * 3 + 1]))
            .then_with(|| coords[a * 3 + 2].cmp(&coords[b * 3 + 2]))
    });

    let mut sorted = Vec::with_capacity(n_atoms * 3);
    for &idx in &indices {
        sorted.push(coords[idx * 3]);
        sorted.push(coords[idx * 3 + 1]);
        sorted.push(coords[idx * 3 + 2]);
    }
    sorted
}

/// Apply origin shifts and atom sorting, returning completed variant tuples.
///
/// For each of the first `num_origin_atoms` atoms, shift so that atom becomes
/// the origin, sort by (label, x, y, z), extract MNF (skip origin), and
/// concatenate with permuted vonorms.
fn apply_shifts_and_sort(
    permuted_vonorms: &[i32],
    transformed_coords: &[i32],
    atom_labels: &[usize],
    n_atoms: usize,
    num_origin_atoms: usize,
    delta: i32,
) -> Vec<Vec<i32>> {
    let mut variants = Vec::with_capacity(num_origin_atoms);
    let motif_len = (n_atoms - 1) * 3;

    for origin_idx in 0..num_origin_atoms {
        let off = origin_idx * 3;
        let sx = -transformed_coords[off];
        let sy = -transformed_coords[off + 1];
        let sz = -transformed_coords[off + 2];

        let mut shifted = vec![0i32; n_atoms * 3];
        for j in 0..n_atoms {
            let jo = j * 3;
            shifted[jo]     = pos_mod(transformed_coords[jo]     + sx, delta);
            shifted[jo + 1] = pos_mod(transformed_coords[jo + 1] + sy, delta);
            shifted[jo + 2] = pos_mod(transformed_coords[jo + 2] + sz, delta);
        }

        let sorted = sort_atoms(&shifted, atom_labels, n_atoms);

        // Build variant: vonorms(7) ++ motif_coords (skip first atom = origin)
        let mut variant = Vec::with_capacity(7 + motif_len);
        variant.extend_from_slice(permuted_vonorms);
        variant.extend_from_slice(&sorted[3..]); // skip origin
        variants.push(variant);
    }

    variants
}


// ---------------------------------------------------------------------------
// Variant generation
// ---------------------------------------------------------------------------

/// Generate all unimodular variants for a single goal CNF.
fn generate_variants_for_goal(
    vonorms: &[i32],
    motif_coords: &[i32],
    elements: &[String],
    delta: i32,
    mode: HeuristicMode,
) -> Vec<Vec<i32>> {
    let atom_labels = compute_atom_labels(elements);
    let n_atoms = elements.len();
    let num_origin_atoms = atom_labels.iter().filter(|&&l| l == 0).count();

    // Reconstruct full coords: origin at (0,0,0) + stored motif coords
    let mut full_coords = vec![0i32; n_atoms * 3];
    full_coords[3..].copy_from_slice(motif_coords);

    let mut variants: Vec<Vec<i32>> = Vec::new();

    match mode {
        HeuristicMode::UnimodularLight => {
            // Iterate over all 168 vonorm permutations (no matrix dedup since
            // different vonorm perms with the same matrix give different permuted
            // vonorms, producing distinct variants).
            for vonorm_perm in PERMUTATIONS.conorm_to_vonorm_perm.values() {
                let permuted_v: Vec<i32> = vonorm_perm.iter().map(|&i| vonorms[i]).collect();

                let idxs = [vonorm_perm[1], vonorm_perm[2], vonorm_perm[3]];
                let u_mat = matrix_from_voronoi_idxs(idxs);
                let u_inv = mat_inv(&u_mat);

                let transformed = transform_coords(&u_inv, &full_coords, n_atoms, delta);
                let new_variants = apply_shifts_and_sort(
                    &permuted_v,
                    &transformed,
                    &atom_labels,
                    n_atoms,
                    num_origin_atoms,
                    delta,
                );
                variants.extend(new_variants);
            }
        }

        HeuristicMode::UnimodularPartial | HeuristicMode::UnimodularFull => {
            let partial = matches!(mode, HeuristicMode::UnimodularPartial);
            let mut seen_mats: HashSet<[i32; 9]> = HashSet::new();

            for (_zero_set, perm_to_mats) in &PERMUTATIONS.zero_to_perm_to_mats {
                for (conorm_perm, mat_list) in perm_to_mats {
                    let vonorm_perm = match PERMUTATIONS.conorm_to_vonorm_perm.get(conorm_perm) {
                        Some(vp) => vp,
                        None => continue,
                    };
                    let permuted_v: Vec<i32> =
                        vonorm_perm.iter().map(|&i| vonorms[i]).collect();

                    let mats_to_use = if partial { &mat_list[..1] } else { &mat_list[..] };

                    for mat_vecs in mats_to_use {
                        let u_mat = [
                            [mat_vecs[0][0], mat_vecs[0][1], mat_vecs[0][2]],
                            [mat_vecs[1][0], mat_vecs[1][1], mat_vecs[1][2]],
                            [mat_vecs[2][0], mat_vecs[2][1], mat_vecs[2][2]],
                        ];

                        if !seen_mats.insert(mat_to_flat(&u_mat)) {
                            continue;
                        }

                        let u_inv = mat_inv(&u_mat);
                        let transformed =
                            transform_coords(&u_inv, &full_coords, n_atoms, delta);
                        let new_variants = apply_shifts_and_sort(
                            &permuted_v,
                            &transformed,
                            &atom_labels,
                            n_atoms,
                            num_origin_atoms,
                            delta,
                        );
                        variants.extend(new_variants);
                    }
                }
            }
        }

        HeuristicMode::Manhattan => unreachable!(),
    }

    // Deduplicate variants
    let mut seen: HashSet<Vec<i32>> = HashSet::new();
    variants.retain(|v| seen.insert(v.clone()));
    variants
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Pre-compute all goal variants for the unimodular manhattan heuristic.
///
/// Called once at the start of A* search. The returned `GoalVariants` is then
/// passed to `unimodular_heuristic()` at every node expansion.
pub fn precompute_goal_variants(
    goal_points: &[(Vec<i32>, Vec<i32>)],
    elements: &[String],
    delta: i32,
    mode: HeuristicMode,
    weight: f64,
) -> GoalVariants {
    let mut all_variants = Vec::new();

    for (vonorms, coords) in goal_points {
        let gv = generate_variants_for_goal(vonorms, coords, elements, delta, mode);
        all_variants.extend(gv);
    }

    if !all_variants.is_empty() {
        eprintln!(
            "Precomputed {} goal variants (mode: {:?}, weight: {:.2})",
            all_variants.len(),
            mode,
            weight,
        );
    }

    GoalVariants {
        variants: all_variants,
        weight,
    }
}

/// Compute the unimodular manhattan heuristic value for a single node.
///
/// Returns `min_over_variants(L1_distance) * weight`.
pub fn unimodular_heuristic(
    vonorms: &[i32],
    coords: &[i32],
    goal_variants: &GoalVariants,
) -> f64 {
    let tuple_len = vonorms.len() + coords.len();
    let mut min_dist = i64::MAX;

    for variant in &goal_variants.variants {
        debug_assert_eq!(variant.len(), tuple_len);

        let mut dist: i64 = 0;
        // Vonorms part
        for (i, &v) in vonorms.iter().enumerate() {
            dist += (v as i64 - variant[i] as i64).abs();
        }
        // Coords part
        let offset = vonorms.len();
        for (i, &c) in coords.iter().enumerate() {
            dist += (c as i64 - variant[offset + i] as i64).abs();
        }

        if dist < min_dist {
            min_dist = dist;
        }
    }

    min_dist as f64 * goal_variants.weight
}
